streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -701,7 +706,15 @@ private StandbyTask convertActiveToStandby(final StreamTask activeTask, final Se
     }
 
     private StreamTask convertStandbyToActive(final StandbyTask standbyTask, final Set<TopicPartition> partitions) {
-        return activeTaskCreator.createActiveTaskFromStandby(standbyTask, partitions, mainConsumer);
+        final StreamTask streamTask = activeTaskCreator.createActiveTaskFromStandby(standbyTask, partitions, mainConsumer);
+        final ProcessorStateManager stateManager = standbyTask.stateManager();
+        for (final TopicPartition partition : partitions) {
+            final ProcessorStateManager.StateStoreMetadata storeMetadata = stateManager.storeMetadata(partition);
+            if (storeMetadata != null && storeMetadata.endOffset() != null) {
+                standbyTaskUpdateListener.onUpdateSuspended(partition, storeMetadata.store().name(), storeMetadata.offset(), storeMetadata.endOffset(), StandbyUpdateListener.SuspendReason.PROMOTED);
```

Comment:
Seems like this is the only place where we invoke the `onUpdateSuspended` callback, ie we're missing the `SuspendReason.MIGRATED` case right?

We can just look to the active restore listener and follow that same example -- looks like it invokes the analogous `#onRestoreSuspended` callback in the StoreChangelogReader, specifically in the `#unregister` method. Personally, I think that would be the best place to invoke the standby listener's `#onUpdateSuspended`, not just in the `MIGRATED` case but also for `PROMOTED`, so we can keep the logic in one place. And `#unregister` is the perfect place to do so, since it always gets invoked whether the task is being closed or recycled. You can just add a parameter to the `#unregister` method to pass in which of those two options it was.

Keeping everything in the StoreChangelogReader also helps us avoid some gnarly questions about special case handling, because the question of when a task is closed can actually be pretty complicated when you look at it within the TaskManager: for example corrupted tasks may be closed and revived, opening up questions about whether and when to invoke `#onUpdateSuspend` if it's going to be revived again. But within the StoreChangelogReader we know that each task should have a 1:1 ratio of calls to `prepareChangelogs` and `unregister`, so it's much easier to reason about

Reply:
Thanks! I agree. I'm going to implement this in the `StoreChangelogReader` adding a new argument for the `#unregister ` method and cover the case for `MIGRATED`

Reply:
I have a theory. We can figure out the `SuspendReason` for `#unregister` method using `ChangelogState`.

When method `#unregister` is called, `StoreChangelogReader` can consider a standby task to be `MIGRATED` when the task's `ChangelogState` is in the `RESTORING` state and PROMOTED when it is in `REGISTERED` state. Since it can be inferred in this instance, we don't need to provide a new argument for the `#unregister` method.

This is based on what I discover from the code, I may be missing something, so I should probably double check this.

Reply:
I think that makes sense -- it's certainly worth a shot! You can always try it out and see if it works 😄 

On that note, there is an existing integration tests that cover task recycling which you might find useful. You can just add a standby listener to that instead of creating new integration tests from scratch. Check out `RestoreIntegrationTest#shouldRecycleStateFromStandbyTaskPromotedToActiveTaskAndNotRestore`

Reply:
I made some adjustments so that we can tell when a 'Task' has been `Migrated` or `Promoted` by looking at its 'ChangelogState' and 'Task' states. Basically, in this way:

Task State: `Running` and ChangelogState: `Restoring` => Promoted
Task State: `Suspended` and ChangelogState: `Restoring` => Migrated

This allows the 'StoreChangelogReader' to infer the cause of the suspension. This seems to be true only for state updater, for some reason, when the state updater is disabled, Task State is always `Suspended` when `#unregister` gets called.

Reply:
I found `RestoreIntegrationTest#shouldRecycleStateFromStandbyTaskPromotedToActiveTaskAndNotRestore` test very useful to understand how all this works, Thanks for the suggestion.
Question: Is there any similar test for migrated standby tasks?